Translation of Apollo Official Chinese Document(s) (#4322)
apolloconfig/apollo: docs/en/usage/apollo-user-guide.md

# II. Public component access guide

## 2.1 Difference between public components and common applications

Public components are those client code that are published for use by other applications, such as the CAT client, Hermes Producer client, etc.

Although such components are developed and maintained by other teams, the runtime is within the actual business application, so they can essentially be considered part of the application.

Usually, the configuration used by these components is maintained by the original development team, but since the runtime and environment of the actual application vary, we also allow the application to override some of the configuration of the public components when they are actually used.

## 2.2 Steps to access public components

The access steps for public components are almost identical to those for normal applications, the only difference being that public components need to create their own unique Namespace.

So, first perform the following steps in the common application access document, and then follow the steps later in this section.

> 1. [Create Project](#_11-%E5%88%9B%E5%BB%BA%E9%A1%B9%E7%9B%AE)

Reviewer: ```suggestion
1. [Create Project](#_11-creating-a-project)
```

Author: fixed.

> 2. [Project administrator privileges](#_121-%E9%A1%B9%E7%9B%AE%E7%AE%A1%E7%90%86%E5%91%98%E6%9D%83%E9%99%90)

Reviewer: ```suggestion
2. [Project administrator privileges](#_121-project-administrator-privileges
```

Author: fixed.

      - ​	![namespace-permission-edit](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/namespace-permission-edit.png)

   2. Assign permission to publish
      - ​	![namespace-publish-permission](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/namespace-publish-permission.png)

6. Click "Back" to return to the project page

### 2.2.2 Adding configuration items

To edit the configuration, you need to have the edit permission of this Namespace. If you find that there is no Add Configuration button, you can find the project administrator to authorize it.

#### 2.2.2.1 Adding configuration via form mode

1. Click Add Configuration
   ![public-namespace-edit-item-entry](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/public-namespace-edit-item-entry.png)

2. Enter the configuration items
   ![public-namespace-edit-item](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/public-namespace-edit-item.png)

3. Click Submit
   ![public-namespace-item-created](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/public-namespace-item-created.png)

#### 2.2.2.2 Edit by text mode

> This part is the same as normal application, please refer to [1.3.2 Editing via text mode](#_132-%E9%80%9A%E8%BF%87%E6%96%87%E6%9C%AC%E6%A8%A1%E5%BC%8F%E7%BC%96%E8%BE%91) for the detailed steps.

Reviewer: ```suggestion
This part is the same as normal application, please refer to [1.3.2 Editing via text mode](#_132-editing-via-text-mode) for the detailed steps.
```

4. With the above configuration, the application deployed in the SHAJQ server room will read the configuration under the SHAJQ cluster

5. If the application is also deployed in other server rooms, then with the above configuration, the configuration under the default cluster will be read.

# IV. Using the same configuration for multiple AppId

In some cases, although the application itself is not a public component, it still needs to share the same configuration among multiple AppId, such as different projects of the same product: XX-Web, XX-Service, XX-Job and so on.

In this case, if you want to implement multiple AppId to use the same configuration, the basic concept is the same as the configuration of public components.

Specifically, we create a namespace under one of the AppId, write the public configuration information, and then read the configuration of the namespace in each project.

If an AppId needs to override the public configuration information, then associate a public namespace under that AppId and write the configuration that needs to be overridden.

> The specific steps can be referred to [Public Component Access Guide](#%e4%ba%8c%e3%80%81%e5%85%ac%e5%85%b1%e7%bb%84%e4%bb%b6%e6%8e%a5%e5%85%a5%e6%8c%87%e5%8d%97).

Reviewer: ```suggestion
The specific steps can be referred to [Public Component Access Guide](#ii-public-component-access-guide).
```

![gray-release-ip-selected](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/gray-release/gray-release-ip-selected.png)

In addition to the IP dimension, from version 2.0.0 onwards there is also support for identifying the list of instances in grayscale by label, which is suitable for scenarios where the IP is not fixed such as `Kubernetes`.

Manually enter the label you want to set, and click the Add button when you're done.

![manual-input-gray-release-label](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/gray-release/manual-input-gray-release-label.png)

![manual-input-gray-release-label-2](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/gray-release/manual-input-gray-release-label2.png)

![gray-release-rule-saved](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/gray-release/gray-release-rule-saved.png)

After the above rules are configured, the grayed configuration will take effect for instances with AppId `100004458`, IP `10.32.21.22` or `Label` marked as `myLabel` or `appLabel`.

> > For more information on how to label `Label`, you can refer to the configuration instructions in [ApolloLabel](en/usage/java-sdk-user-guide?id=_1247-apollolabel).

Reviewer: ```suggestion
> For more information on how to label `Label`, you can refer to the configuration instructions in [ApolloLabel](en/usage/java-sdk-user-guide?id=_1247-apollo-label).
```

> If the required IP is not found in the drop-down box, it means that the machine has not yet taken the configuration from Apollo, you can enter it by clicking Enter IP manually, and click the Add button after entering it
> 
> After entering the IP, click the Add button. [manual-input-gray-release-ip](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/gray-release/manual-input-gray-release-ip.png)

Reviewer: ```suggestion
If the required IP is not found in the drop-down box, it means that the machine has not yet taken the configuration from Apollo, you can enter it by clicking Enter IP manually, and click the Add button after entering the IP.

![manual-input-gray-release-ip](https://cdn.jsdelivr.net/gh/apolloconfig/apollo@master/doc/images/gray-release/manual-input-gray-release-ip.png)
```

The configuration is very simple. After logging in with your super administrator account, go to the `Administrator Tools - System Parameters` page and add or modify the `configView.memberOnly.envs` configuration item.

![configView.memberOnly.envs](https://user-images.githubusercontent.com/837658/46456519-c155e100-c7e1-11e8-969b-8f332379fa29.png)

## 6.2 Configuring access keys

Apollo has added an access key mechanism since version 1.6.0, so that only authenticated clients can access sensitive configurations. If the application has access keys enabled, the client needs to configure the keys, otherwise the configuration cannot be accessed.

1. The project administrator opens the Manage Keys page
   ![Admin Key Portal](https://user-images.githubusercontent.com/837658/94990081-f4d3cd80-05ab-11eb-9470-fed5ec6de92e.png)

2. Generate an access key for each environment of the project, note that it is disabled by default, and it is recommended to turn it on after the clients are all configured
   ![Key Configuration Page](https://user-images.githubusercontent.com/837658/94990150-788dba00-05ac-11eb-9a12-727fdb872e42.png)

> 3. Client-side [configure access key](en/usage/java-sdk-user-guide#_1244-configure-access-key)

Reviewer: ```suggestion
3. Client-side [configure access key](en/usage/java-sdk-user-guide#_1244-configuring-access-keys)
```

# VII. Best practices

## 7.1 Security-related

As a basic service, the configuration center stores very important configuration information of the company, so security factors need to be focused on, some considerations are listed below for your reference, and you are also welcome to share your own practice cases.

### 7.1.1 Authentication

It is recommended to access the company's unified authentication system, such as SSO, LDAP, etc. The access method can be found in [Portal to implement user login function](en/development/portal-how-to-implement-user-login-function)

> If you use Spring Security simple authentication provided by Apollo, you must remember to change the password of super administrator apollo

### 7.1.2 Authorization

Apollo supports fine-grained permissions control, so please make sure to control the permissions according to the actual situation: 1.

1. [Project administrator privileges](#_121-Project-administrator-privileges)

>    * Apollo allows all logged-in users to create projects by default. If you only allow some users to create projects, you can turn on [Create project permission control](en/deployment/distributed-deployment-guide#_3110-role.create-application.enabled---whether-to-enable-control-over-creating-project-permissions)

Reviewer: ```suggestion
   * Apollo allows all logged-in users to create projects by default. If you only allow some users to create projects, you can turn on [Create project permission control](en/deployment/distributed-deployment-guide#_3110-rolecreate-applicationenabled-whether-to-enable-control-over-creating-project-permissions)
```

2. [Configure editing and publishing privileges](#_122-Configuring-editing-and-publishing-permissions)
   * Configuration editing and publishing privileges support configuration by environment, for example, the development environment developers can complete the process of configuration editing and publishing by themselves, but the production environment publishing privileges to the test or operation and maintenance personnel

>    * It is recommended to turn on [release audit](en/deployment/distributed-deployment-guide#_322-namespace.lock.switch---Only-one-person-can-modify-the-switch-at-a-time-for-release-review) at the same time for production environment, so as to control that only one person can modify a configuration release and another person can release it. This ensures that configuration changes are adequately checked.

Reviewer: ```suggestion
   * It is recommended to turn on [release audit](en/deployment/distributed-deployment-guide#_322-namespacelockswitch-only-one-person-can-modify-the-switch-at-a-time-for-release-review) at the same time for production environment, so as to control that only one person can modify a configuration release and another person can release it. This ensures that configuration changes are adequately checked.
```

3. [Configuration view permissions](#_61-Configure-view-permissions)
   * You can specify that only project members of an environment are allowed to view the configuration of a private Namespace, thus avoiding sensitive configuration leaks, such as production environments

### 7.1.3 System access

In addition to user permissions, system access also needs to be considered in terms of.

1. `apollo-configservice` and `apollo-adminservice` are designed based on the intranet trusted network, so for security reasons, `apollo-configservice` and `apollo-adminservice` are prohibited from being exposed directly to the public network
2. For sensitive configurations, consider enabling [access secret key](#_62-Configuring-access-keys) so that only authenticated clients can access sensitive configurations

> 3. 1.7.1 and above can consider enabling [access control](en/deployment/distributed-deployment-guide?#_326-admin-services.access.control.enabled---Configure-whether-apollo-adminservice-has-access-control-enabled) for `apollo-adminservice`, so that only [controlled](en/deployment/distributed-deployment-guide#_3112-admin-services.access.token---set-the-access-token-required-by-apollo-portal-to-access-the-apollo-adminservice-for-each-environment) `apollo-portal` can access the corresponding interface to enhance security

Reviewer: ```suggestion
3. 1.7.1 and above can consider enabling [access control](en/deployment/distributed-deployment-guide?#_326-admin-servicesaccesscontrolenabled-configure-whether-apollo-adminservice-has-access-control-enabled) for `apollo-adminservice`, so that only [controlled](en/deployment/distributed-deployment-guide#_3112-admin-servicesaccesstokens-set-the-access-token-required-by-apollo-portal-to-access-the-apollo-adminservice-for-each-environment) `apollo-portal` can access the corresponding interface to enhance security
```
